lib/services/services.c

```diff
@@ -1079,3 +1080,82 @@ resources_list_agents(const char *standard, const char *provider)
 
     return NULL;
 }
+
+gboolean
+resources_valid_tuple(const char *standard, const char *provider, const char *agent)
+{
+    GList *standards = NULL;
+    GList *providers = NULL;
+    GList *agents = NULL;
+    GListPtr iter = NULL;
+    gboolean rc = FALSE;
+
+    if (crm_str_eq(standard, "stonith", TRUE)) {
+        stonith_t *st = NULL;
+        stonith_key_value_t *devices = NULL;
+        stonith_key_value_t *dIter = NULL;
+        int timeout = 120;
+
+        if (provider != NULL || agent == NULL) {
+            return rc;
+        }
+
+        st = stonith_api_new();
+        st->cmds->list_agents(st, st_opt_sync_call, NULL, &devices, timeout);
+
+        for (dIter = devices; dIter != NULL; dIter = dIter->next) {
+            if (crm_str_eq(dIter->value, agent, TRUE)) {
+                rc = TRUE;
+                break;
+            }
+        }
+
+        stonith_key_value_freeall(devices, 1, 1);
+        stonith_api_delete(st);
+        return rc;
+    }
```

Comment:
This would require anything linked against libcrmservice to also link against libstonithd, which would be a break in backward API/ABI compatibility. I think we have to follow the current model of separating the stonith portion into libstonithd and have just the non-stonith portion here, with the caller responsible for calling the correct one.

Alternatively, we could put this in a private library for pacemaker's internal use only that then has to be linked against libcrmservice and libstonith. There is an existing libpengine (confusingly created via daemons/schedulerd/Makefile.am, consisting of everything except pacemaker-schedulerd.c) that could be broadened into such a thing.

Reply:
The code for figuring out whether a stonith agent exists is so different that splitting it into two different functions and making the caller responsible seems reasonable.